psydac/api/ast/fem.py

```diff
@@ -246,11 +246,20 @@ class Kernel(SplBasic):
 
     def __new__(cls, weak_form, kernel_expr, target=None,
                 discrete_boundary=None, name=None, boundary_basis=None,
-                mapping=None, is_rational_mapping=None,backend=None):
+                mapping=None, is_rational_mapping=None,symbolic_space=None, backend=None):
 
         if not isinstance(weak_form, FunctionalForms):
             raise TypeError('> Expecting a weak formulation')
 
+        if symbolic_space:
+            symbolic_space= symbolic_space[0]
+            
+        is_compatible_spaces = False
+        if isinstance(symbolic_space, ProductSpace):
+            spaces = symbolic_space.spaces
+            space = spaces[0]
+            is_compatible_spaces = not all(sp.kind==space.kind for sp in spaces)
```

Comment:
no it's not enough of course Sympde checks that when we construct the Bilinear or the Linear form, in fact is_compatible_spaces = is_different_spaces which means the spaces are not identical because in the beginning we were assuming  that they were identical and the code genation is different in these two cases

Reply:
I see. So this flag only makes sense if we are working on a **ProductSpace**, and tells us that the component spaces are not all identical. For example if the 2D velocity vector field `u` belongs to a certain space, its components `ux` and `uy` may belong to different (compatible) spaces. Is this correct?

Reply:
yes exactly

Reply:
OK then I have the following question. For the example that I have just given, the space of `u` is usually described in SymPDE by the concept of `VectorFunctionSpace`, while the `ProductSpace` is used for the space of a tuple `(u, p)` for example. Is this going to be different if we use compatible spaces?

Reply:
Please forget the last comment. For completeness I think we should also check the case where `symbolic_space` is a **VectorFunctionSpace**.

Reply:
okay it's done